### DISCUSSION TITLE
How to pass dynamic args to useContractRead

### DISCUSSION THREAD

**Discussion by ujjwala2296** [wevm/wagmi#449] [Q&A] [answered]
I am able to pass args to useContractWrite but not useContractRead

**Comment by jxom** [comment, marked as answer]
Can you please provide a code snippet?

**Reply by ujjwala2296** [reply]
I want to read contract in a for loop, with dynamic args from an array.

`for (const position of data.Positions) {
	const startPriceA = await priceFeed
	.connect(provider)
	.getPrice(position.tokenA, position.startEpoch);

    const startPriceB = await priceFeed
      .connect(provider)
      .getPrice(position.tokenB, position.startEpoch);
}`

I want to change this and use useContractRead hook

**Reply by jxom** [reply]
Are you trying to read from a contract directly? The `useContractRead` hook is only compatible with deployed smart contracts. The example you provided looks like it is querying some sort of SDK, but I'm assuming it is querying a smart contract underneath? If so, I'd suggest:

1. Obtaining the smart contract address,

2. Retrieve & download the ABI from Etherscan (e.g. https://api.etherscan.io/api?module=contract&action=getabi&address=0xecb504d39723b0be0e3a9aa33d646642d1051ee1 – replace the address with yours)

3. Import the ABI and `useContractRead` in your React code

```tsx
import { useContractRead } from 'wagmi';
import wagmigotchiAbi from './wagmigotchiAbi.json';

function App() {
  const { data, isError, isLoading } = useContractRead(
    {
      addressOrName: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
      contractInterface: wagmigotchiAbi,
    },
    'getHunger',
  )
}
```

**Reply by ujjwala2296** [reply]
I am using deployed contract, I am importing the below code and using it.

` const priceFeed = new ethers.Contract(
      PriceFeedDeployment.address,
      PriceFeedDeployment.abi
    );`

**Reply by jxom** [reply]
Ah right.

`useContractRead` only supports defining arguments in the hook, and not via a fetcher like `useContractWrite` does. If you want to use an array of args to invoke your contract, you can:

1. Place `useContractRead` in a component that is rendered within the list:

```tsx
import { useContractRead } from 'wagmi'
import { PriceFeedDeployment } from './somewhere'

function Example({ positions }) {
  return (
    <>
      {positions.map(({ tokenA, tokenB, startEpoch }, i) => 
        <Position 
          key={i} 
          tokenA={tokenA} 
          tokenB={tokenB}
          startEpoch={startEpoch}
        />
      )}
    </>
  )
}

function Position({ tokenA, tokenB, startEpoch }) {
  const { data: priceA } = useContractRead(
    {
      addressOrName: PriceFeedDeployment.address,
      contractInterface: PriceFeedDeployment.abi,
    },
    'getPrice',
    { args: [tokenA, startEpoch] }
  )
  const { data: priceB } = useContractRead(
    {
      addressOrName: PriceFeedDeployment.address,
      contractInterface: PriceFeedDeployment.abi,
    },
    'getPrice',
    { args: [tokenB, startEpoch] }
  )
  
  ...
}
```

2. Batch the direct contract calls inside a `useQuery` ([`useQuery` API is identical to react-query](https://react-query.tanstack.com/reference/useQuery)). Note: The `useQuery` hook is used internally by `useContractRead`, and you can also place any async function as a second parameter to `useQuery` (just make sure it returns some data).

```tsx
import { useContract, useQuery } from 'wagmi'
import { PriceFeedDeployment } from './somewhere'

function Example({ positions }) {
  const contract = useContract({
    addressOrName: PriceFeedDeployment.address,
    contractInterface: PriceFeedDeployment.abi,
  })
  
  const { data } = useQuery('prices', async () => {
    let result = []
    // loop through `positions` and call `contract.getPrice` directly in here and append to `result` – similar to the example you provided.
    return result;
  })
}
```

**Reply by ujjwala2296** [reply]
Second Option was really helpful. Thanks.

**Reply by jxom** [reply]
That said, it would be nice if batched contract reads were a first-class API. Also discussed [here](https://github.com/tmm/wagmi/discussions/452#discussioncomment-2742049).

**Reply by salieflewis** [reply]
Is it possible to pass in props to the addressOrName field? 

```
const Purchase = (address, price) => {
  const provider = useProvider(); // Alchemy
  console.log(price);

  const { data, isError, isLoading, write } =
    useContractWrite({
      addressOrName: address,
      contractInterface: ERC721Drop,
      functionName: 'purchase',
      args: [price],
    });

    return <button onClick={() => write()}>mint</button>;

export default Purchase
```

**Reply by davidlin504** [reply]
Sorry for disturbance! I am beginer about wagmi.
is it possible to pass args as the form of args: [[1], [0], ['0x.....']]

**Reply by salieflewis** [reply]
What you wrote would suggest that you're providing array values for each argument.

I'm guessing you would probably want to just be passing in those arguments like this

`args: ['1', '0', '0x...']`

Not actually sure those arguments need to be passed as strings also, I think it is situational.

Perhaps you can get more context via the [docs](https://wagmi.sh/react/prepare-hooks/usePrepareContractWrite), hope that helps!

**Reply by davidlin504** [reply]
Thanks for your replay. @salieflewis  It is my fault that I do not append the abi format
here is the part of my abi inputs 
```
[ {"inputs": [
    {
      "internalType": "uint8[]",
      "name": "actions",
      "type": "uint8[]"
    },
    {
      "internalType": "uint256[]",
      "name": "values",
      "type": "uint256[]"
    },
    {
      "internalType": "bytes[]",
      "name": "datas",
      "type": "bytes[]"
    }
  ],
.....
....
}]
```

I check the code  at node_modules/@wagmi/core/dist/chunk-VRBYTA34.js 
function isArgOfType(arg, abiParameter) {.....} can accept array
 
Now, I tried hard to figure out why did not useContractWrite work properly.

**Reply by davidlin504** [reply]
updated I found a solution
`usePrepareContractWrite`  comes with a params `enabled`
When I set enabled: true, my code works.

**Comment by wishwaD7** [comment]
can I pass an array as an argument in useContrctRead?

**Reply by 1337-ishaan** [reply]
+1
